test/TestApps/PushNotificationsTestApp/main.cpp

```diff
@@ -0,0 +1,423 @@
+    multipleChannelRequestUsingMultipleRemoteId, threeChannelRequestUsingSameRemoteId, activatorTest,
+};
+
+static std::map<std::string, UnitTest> switchMapping;
+winrt::guid remoteId1(L"a2e4a323-b518-4799-9e80-0b37aeb0d225");
+winrt::guid remoteId2(L"CA1A4AB2-AC1D-4EFC-A132-E5A191CA285A");
+winrt::guid remoteId3(L"40FCE789-C6BF-4F47-A6CF-6B9C1DCE31BA");
+
+void signalPhase(const std::wstring& phaseEventName)
+{
+    wil::unique_event phaseEvent;
+    if (phaseEvent.try_open(phaseEventName.c_str(), EVENT_MODIFY_STATE, false))
+    {
+        phaseEvent.SetEvent();
+    }
+}
+
+void initUnitTestMapping()
+{
+    switchMapping["ChannelRequestUsingNullRemoteId"] = UnitTest::channelRequestUsingNullRemoteId;
+    switchMapping["ChannelRequestUsingRemoteId"] = UnitTest::channelRequestUsingRemoteId;
+    switchMapping["MultipleChannelRequestUsingSameRemoteId"] = UnitTest::multipleChannelRequestUsingSameRemoteId;
+    switchMapping["MultipleChannelRequestUsingMultipleRemoteId"] = UnitTest::multipleChannelRequestUsingMultipleRemoteId;
+    switchMapping["ThreeChannelRequestUsingSameRemoteId"] = UnitTest::threeChannelRequestUsingSameRemoteId;
+    switchMapping["ActivatorTest"] = UnitTest::activatorTest;
+}
+
+bool ChannelRequestUsingNullRemoteId()
+{
+    winrt::hresult hr = S_OK;
+
+    try
+    {
+        auto channelOperation = PushNotificationManager::CreateChannelAsync(GUID_NULL).get();
+    }
+    catch (...)
+    {
+        auto channelRequestException = hresult_error(to_hresult());
+        hr = channelRequestException.code();
+    }
+
+    return (hr == E_INVALIDARG);
+}
+
+bool ChannelRequestUsingRemoteId()
+{
+    wil::unique_handle channelEvent = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
+    auto channelOperationResult = S_OK;
+    auto channelOperation = PushNotificationManager::CreateChannelAsync(remoteId1);
+
+    // Setup the completed event handler
+    channelOperation.Completed(
+        [&channelEvent, &channelOperationResult](
+            IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& sender,
+            AsyncStatus const /* asyncStatus */)
+        {
+            auto result = sender.GetResults();
+            if (result.Status() == PushNotificationChannelStatus::CompletedSuccess)
+            {
+                result.Channel().Close(); // Avoids using cached channel everytime the test runs
+            }
+            else if (result.Status() == PushNotificationChannelStatus::CompletedFailure)
+            {
+                channelOperationResult = result.ExtendedError();
+            }
+
+            SetEvent(channelEvent.get());
+        });
+
+    // The maximum amount of time it takes for channel request to be obtained - 16mins
+    if (WAIT_OBJECT_0 != WaitForSingleObject(channelEvent.get(), 960000 /* milliseconds */))
+    {
+        channelOperation.Cancel();
+        channelOperationResult = ERROR_TIMEOUT;
```

Comment:
If the operation times out, then we return immediately even though the Completed handler is going to run eventually and try to use the reference-captured event handles (which are now references to destroyed objects).

In this case, I think the thing to do is not to use hook up to Completed with a custom event but rather do

```cpp
bool ChannelRequestUsingRemoteId()
{
    auto channelOperation = PushNotificationManager::CreateChannelAsync(remoteId1);

    if (channelOperation.wait_for(16min) != AsyncStatus::Completed)
    {
        channelOperation.Cancel();
        return false; // timed out or failed
    }
    auto result = channelOperation.GetResults();
    auto status = result.Status();
    if (status != PushNotificationChannelStatus::CompletedSuccess)
    {
        return false; // did not produce a channel
    }
    result.Channel().Close(); // Avoids using cached channel everytime the test runs
    return true;
}
```